src/Microsoft.VisualStudio.LanguageServices.Razor/DefaultTagHelperCompletionService.cs

```diff
@@ -21,6 +22,99 @@ public DefaultTagHelperCompletionService(TagHelperFactsService tagHelperFactsSer
         }
 
+        public override AttributeCompletionResult GetAttributeCompletions(AttributeCompletionContext completionContext)
+        {
+            if (completionContext == null)
+            {
+                throw new ArgumentNullException(nameof(completionContext));
+            }
+
+            var attributeCompletions = completionContext.ExistingCompletions.ToDictionary(
+                completion => completion,
+                _ => new HashSet<BoundAttributeDescriptor>(),
+                StringComparer.OrdinalIgnoreCase);
+
+            var documentContext = completionContext.DocumentContext;
+            var descriptorsForTag = _tagHelperFactsService.GetTagHelpersGivenTag(documentContext, completionContext.CurrentTagName, completionContext.CurrentParentTagName);
+            if (descriptorsForTag.Count == 0)
+            {
+                // If the current tag has no possible descriptors then we can't have any additional attributes.
+                var defaultResult = AttributeCompletionResult.Create(attributeCompletions);
+                return defaultResult;
+            }
+
+            var prefix = documentContext.Prefix ?? string.Empty;
+            Debug.Assert(completionContext.CurrentTagName.StartsWith(prefix, StringComparison.OrdinalIgnoreCase));
+
+            var applicableTagHelperBinding = _tagHelperFactsService.GetTagHelperBinding(documentContext, completionContext.CurrentTagName, completionContext.Attributes, completionContext.CurrentParentTagName);
+            var applicableDescriptors = applicableTagHelperBinding?.Descriptors ?? Enumerable.Empty<TagHelperDescriptor>();
+            var unprefixedTagName = completionContext.CurrentTagName.Substring(prefix.Length);
+
+            if (!completionContext.InHTMLSchema(unprefixedTagName) &&
+                applicableDescriptors.All(descriptor => descriptor.TagOutputHint == null))
+            {
+                // This isn't a known HTML tag and no descriptor has an output element hint. Remove all previous completions.
+                attributeCompletions.Clear();
+            }
+
+            for (var i = 0; i < descriptorsForTag.Count; i++)
+            {
+                var curDescriptor = descriptorsForTag[i];
+
+                if (applicableDescriptors.Contains(curDescriptor))
+                {
+                    // Add attributes from satisfied bound attribute descriptors.
+                    foreach (var curAttributeDescriptor in curDescriptor.BoundAttributes)
+                    {
+                        UpdateCompletions(curAttributeDescriptor.Name, curAttributeDescriptor);
+                    }
+                }
+                else
+                {
+                    var lookup = curDescriptor.BoundAttributes.ToDictionary(attribute => attribute.Name, StringComparer.OrdinalIgnoreCase);
+
+                    // If requirements aren't satisifed for this taghelper, add RequiredAttributes from all rules.
```

Comment:
Same thing, this says what it does, but not really in enough detail to be meaningful. You could add an example or define some terminology